### ISSUE TITLE
Can't mix tensors and python objects

### ISSUE THREAD

**Issue by vladfi1** [tensorflow/tensorflow#2328]
I'd like to do things such as the following:

```python
>>> tf.convert_to_tensor([tf.constant(1), 2])
>>> tf.slice(..., [0, 1, some_tensor], ...)
```

Currently both of these raise exceptions. The alternative is to use a combination of `tf.expand_dims` and `tf.concat`, which is a bit unwieldy.

**Comment by mrry** [comment]
Hi there,

Several people have requested this recently, and I think it's time we implemented this. I'll take on the feature request.

Derek.

**Comment by rdadolf** [comment]
Just pulling in the conversation from #2318, since it's really more appropriate here.

The downside of this feature is type errors can become harder to hunt down. For instance, having a dimension typo can become a logical error, especially if it's compounded with broadcasting.

While manually converting heterogeneous structures may seem tedious, it's also clearer what's going on. It might seem that `tf.convert_to_tensor([tf.constant(1), 2])` is already pretty clear, but if you add a couple layers of variables in between and end up with `tf.convert_to_tensor([x,y])`, a type error is pretty easy to lose.

My feeling is "explicit is better than implicit" here. Perhaps there's some compromise where we don't do recursive auto-coercion for all things that can take tensors? Maybe just allow a single explicit method to handle this behavior? `convert_to_tensor` *would* seem like a decent candidate, so the above code would become:

```
tf.convert_to_tensor([tf.constant(1), 2]) # works, this is the explicit coercion method
tf.slice(..., tf.convert_to_tensor([0, 1, some_tensor]), ...)
```

**Comment by vladfi1** [comment]
@rdadolf I don't see why type errors become logical errors - they'll just point to the nested object instead of the root. In any case, the current behavior allows both python objects and tensors, just not mixed - I assume this is done with an implicit call to `convert_to_tensor`, so any improvements to `convert_to_tensor` should carry over to all its use-sites.

**Comment by altaetran** [comment]
I am personally not a fan of mixed typing like this. I see tensorflow as a programming language with a python wrapper, rather than python module that can interact nicely with other python modules.

**Comment by rdadolf** [comment]
> I don't see why type errors become logical errors - they'll just point to the nested object instead of the root.

@vladfi1 The type errors I'm talking about are dimension mismatch due to extra or insufficient nesting. Consider the following exceedingly contrived example:

```
x = tf.constant([1,2])
y = tf.constant([1,2])
x = tf.constant(x)
y = tf.constant([y]) # oops
w = tf.constant([[1,2,],[3,4]])
print tf.Session().run( tf.add(w, tf.reduce_max(x,0)) )
print tf.Session().run( tf.add(w, tf.reduce_max(y,0)) )
```

Currently, TF gives you a type error; in the proposed extension, this silently gives you the wrong answer (i.e., it becomes a logical error). This is obviously not real code, but imagine having an n-dim tensor accidentally becoming an n+1 dim tensor whilst threading through a larger body of code, and it can easily end up costing you a day of hunting down the bug.

I'm not suggesting that we put every TF programmer into a padded room to protect them from themselves, but I feel that libraries which try to make everything convertible to everything else end up a mess to debug. This kind of error protection is the reason behind type systems. We should take advantage of it!

>  In any case, the current behavior allows both python objects and tensors, just not mixed - I assume this is done with an implicit call to convert_to_tensor, so any improvements to convert_to_tensor should carry over to all its use-sites.

This isn't actually the case. The error stems from [here](https://github.com/tensorflow/tensorflow/blob/17dcc5a176d152caec570452d28fb94920cceb8c/tensorflow/python/framework/tensor_util.py#L273), which is used by [this](https://github.com/tensorflow/tensorflow/blob/17dcc5a176d152caec570452d28fb94920cceb8c/tensorflow/python/framework/tensor_util.py#L284), which is used directly by ops (e.g., [here](https://github.com/tensorflow/tensorflow/blob/17dcc5a176d152caec570452d28fb94920cceb8c/tensorflow/python/ops/constant_op.py#L162)). Then there is a whole separate set of conversion routines in the [tensor conversion registry](https://github.com/tensorflow/tensorflow/blob/82ff4cd8b0d541ede107d34d8eecc769c91dda11/tensorflow/python/framework/ops.py#L520), which is [also used](https://github.com/tensorflow/tensorflow/blob/82ff4cd8b0d541ede107d34d8eecc769c91dda11/tensorflow/python/framework/ops.py#L569) by `convert_to_tensor`. And as mentioned in the other thread, this is also separate from `tf.Session().run()`'s feed dict conversion registry.
